src/Models/Model.php

```diff
@@ -354,6 +391,49 @@ public function jsonSerialize() {
 		return get_object_vars( $this );
 	}
 
+	/**
+	 * Constructs a model instance from database query data.
+	 *
+	 * @param object|array $queryData
+	 * @param int $mode The level of strictness to take when constructing the object, by default it will ignore extra keys but error on missing keys.
+	 * @return static
+	 */
+	public static function fromData($data, $mode = self::BUILD_MODE_IGNORE_EXTRA) {
+		if ( ! is_object( $data ) && ! is_array( $data ) ) {
+			Config::throwInvalidArgumentException( 'Query data must be an object or array' );
+		}
+
+		$data = (array) $data;
+
+		// If we're not ignoring extra keys, check for them and throw an exception if any are found.
+		if ( ! ($mode & self::BUILD_MODE_IGNORE_EXTRA) ) {
+			$extraKeys = array_diff_key( (array) $data, static::$properties );
+			if ( ! empty( $extraKeys ) ) {
+				Config::throwInvalidArgumentException( 'Query data contains extra keys: ' . implode( ', ', array_keys( $extraKeys ) ) );
+			}
+		}
+
+		if ( ! ($mode & self::BUILD_MODE_IGNORE_MISSING) ) {
+			$missingKeys = array_diff_key( static::$properties, (array) $data );
+			if ( ! empty( $missingKeys ) ) {
+				Config::throwInvalidArgumentException( 'Query data is missing keys: ' . implode( ', ', array_keys( $missingKeys ) ) );
+			}
+		}
+
+		$instance = new static();
```

Comment:
This could resolve in errors without a `final` constructor. Then again the library has to allow developers to make changes during each model's construction...

I would go with a `final` constructor to ensure the signature of every possible constructor remains the same, then introduce a protected method that will allow developers extending the abstract to utilize for during construction functionality.

Reply:
Good catch! Resolved in f69abd6